benchbuild/project.py

```diff
@@ -154,69 +165,54 @@ class Project(metaclass=ProjectDecorator):
-        """Create a new project instance and set some defaults."""
-        new_self = super(Project, cls).__new__(cls)
-        if cls.NAME is None:
-            raise AttributeError(
-                "{0} @ {1} does not define a NAME class attribute.".format(
-                    cls.__name__, cls.__module__))
-        if cls.DOMAIN is None:
-            raise AttributeError(
-                "{0} @ {1} does not define a DOMAIN class attribute.".format(
-                    cls.__name__, cls.__module__))
-        if cls.GROUP is None:
-            raise AttributeError(
-                "{0} @ {1} does not define a GROUP class attribute.".format(
-                    cls.__name__, cls.__module__))
-        return new_self
-
-    experiment = attr.ib()
-
-    name = attr.ib(
-        default=attr.Factory(lambda self: type(self).NAME, takes_self=True))
-
-    domain = attr.ib(
-        default=attr.Factory(lambda self: type(self).DOMAIN, takes_self=True))
-
-    group = attr.ib(
-        default=attr.Factory(lambda self: type(self).GROUP, takes_self=True))
-
-    src_file = attr.ib(
+    NAME: str = None
+    DOMAIN: str = None
+    GROUP: str = None
+    VERSION: str = None
+    SRC_FILE: str = None
+    CONTAINER: Container = None
+
+    experiment: Experiment = attr.ib()
+    name: str = attr.ib(
+        default=attr.Factory(lambda self: type(self).NAME, takes_self=True),
+        validator=validators.attribute_not_empty('NAME'))
+
+    domain: str = attr.ib(
+        default=attr.Factory(lambda self: type(self).DOMAIN, takes_self=True),
+        validator=validators.attribute_not_empty('DOMAIN'))
+
+    group: str = attr.ib(
+        default=attr.Factory(lambda self: type(self).GROUP, takes_self=True),
+        validator=validators.attribute_not_empty('GROUP'))
+
+    src_file: str = attr.ib(
         default=attr.Factory(
             lambda self: type(self).SRC_FILE, takes_self=True))
 
-    container = attr.ib(
+    container: Container = attr.ib(
         default=attr.Factory(
             lambda self: type(self).CONTAINER, takes_self=True))
 
-    version = attr.ib(
+    version: str = attr.ib(
         default=attr.Factory(lambda self: type(self).VERSION, takes_self=True))
 
-    testdir = attr.ib()
+    testdir: LocalPath = attr.ib()
 
     @testdir.default
-    def __default_testdir(self):
+    def __default_testdir(self) -> LocalPath:
         if self.group:
             return local.path(str(
                 CFG["test_dir"])) / self.domain / self.group / self.name
         return local.path(str(CFG["test_dir"])) / self.domain / self.name
 
-    cflags = attr.ib(default=attr.Factory(list))
+    cflags: List[str] = attr.ib(default=attr.Factory(list))
 
-    ldflags = attr.ib(default=attr.Factory(list))
+    ldflags: List[str] = attr.ib(default=attr.Factory(list))
 
-    run_uuid = attr.ib()
+    run_uuid: uuid.UUID = attr.ib()
 
     @run_uuid.default
-    def __default_run_uuid(self):
+    def __default_run_uuid(self) -> uuid.UUID:
```

Comment:
R0201 (no-self-use): Method could be a function  <!-- Automatically posted by Lintly -->